In [1]:
from nba_api.stats.static import players
from nba_api.stats.endpoints import PlayerGameLog
import pandas as pd
import nba_api.stats.endpoints
import time
import random

In [2]:
draft_df = nba_api.stats.endpoints.drafthistory.DraftHistory().get_data_frames()[0]

In [3]:
draft_2003 = draft_df[draft_df["SEASON"] == "2003"]
draft_2003

PERSON_ID             PLAYER_NAME SEASON  ROUND_NUMBER  ROUND_PICK  \
1312       2544            LeBron James   2003             1           1   
1313       2545           Darko Milicic   2003             1           2   
1314       2546         Carmelo Anthony   2003             1           3   
1315       2547              Chris Bosh   2003             1           4   
1316       2548             Dwyane Wade   2003             1           5   
1317       2549             Chris Kaman   2003             1           6   
1318       2550            Kirk Hinrich   2003             1           7   
1319       2551               T.J. Ford   2003             1           8   
1320       2552        Michael Sweetney   2003             1           9   
1321       2553            Jarvis Hayes   2003             1          10   
1322       2554         Mickael Pietrus   2003             1          11   
1323       2555           Nick Collison   2003             1          12   
1324       2556            Marcus Banks   2003             1          13   
1325       2557            Luke Ridnour   2003             1          14   
1326       2558            Reece Gaines   2003             1          15   
1327       2559               Troy Bell   2003             1          16   
1328       2560         Zarko Cabarkapa   2003             1          17   
1329       2561              David West   2003             1          18   
1330       2562          Sasha Pavlovic   2003             1          19   
1331       2563           Dahntay Jones   2003             1          20   
1332       2564              Boris Diaw   2003             1          21   
1333       2565          Zoran Planinic   2003             1          22   
1334       2566           Travis Outlaw   2003             1          23   
1335       2567              Brian Cook   2003             1          24   
1336       2568          Carlos Delfino   2003             1          25   
1337       2569               Ndudi Ebi   2003             1          26   
1338       2570        Kendrick Perkins   2003             1          27   
1339       2571         Leandro Barbosa   2003             1          28   
1340       2572             Josh Howard   2003             1          29   
1341       2573            Maciej Lampe   2003             2           1   
1342       2574            Jason Kapono   2003             2           2   
1343       2575             Luke Walton   2003             2           3   
1344       2602          Jerome Beasley   2003             2           4   
1345       2577  Sofoklis Schortsanitis   2003             2           5   
1346       2578         Szymon Szewczyk   2003             2           6   
1347       2579            Mario Austin   2003             2           7   
1348       2580           Travis Hansen   2003             2           8   
1349       2581             Steve Blake   2003             2           9   
1350       2582           Slavko Vranes   2003             2          10   
1351       2583       Derrick Zimmerman   2003             2          11   
1352       2584            Willie Green   2003             2          12   
1353       2585           Zaza Pachulia   2003             2          13   
1354       2586            Keith Bogans   2003             2          14   
1355       2587          Malick Badiane   2003             2          15   
1356       2588             Matt Bonner   2003             2          16   
1357       2589          Sani Becirovic   2003             2          17   
1358       2590             Mo Williams   2003             2          18   
1359       2591              James Lang   2003             2          19   
1360       2592             James Jones   2003             2          20   
1361       2593       Paccelis Morlende   2003             2          21   
1362       2594             Kyle Korver   2003             2          22   
1363       2595       Remon Van de Hare   2003        

In [4]:
players_name_id = draft_2003[["PLAYER_NAME", "PERSON_ID"]]

In [5]:
players_10_games = []
for player_id, player_name in zip(players_name_id["PERSON_ID"], players_name_id["PLAYER_NAME"]):
	game_log = PlayerGameLog(player_id=player_id, season="2003-04", season_type_all_star="Regular Season")
	df = game_log.get_data_frames()[0]
	df["GAME_DATE"] = pd.to_datetime(df["GAME_DATE"])
	df_sorted = df.sort_values("GAME_DATE")
	first_10_games = df_sorted.head(10)
	if not first_10_games.empty:	# Ensuring they have played at least a game
		player_stats = {
			"player_name": player_name,
			"player_id": player_id,
			"fta": first_10_games["FTA"].sum(),
			"ftm": first_10_games["FTM"].sum(),
			"ft_pct": first_10_games["FTM"].sum() / first_10_games["FTA"].sum() if first_10_games["FTA"].sum() > 0 else 0
		}
		players_10_games.append(player_stats)
	time.sleep(random.uniform(0.5, 1.5))  # To avoid hitting rate limits
df_players_10_games = pd.DataFrame(players_10_games)

In [6]:
# removing players who didnt shoot free throws in their first 10 games
# since they dont provide meaningful data for free throw percentage
df_players_10_games = df_players_10_games[df_players_10_games["fta"] > 0]
df_players_10_games

player_name  player_id  fta  ftm    ft_pct
0       LeBron James       2544   38   23  0.605263
1      Darko Milicic       2545    3    1  0.333333
2    Carmelo Anthony       2546   60   43  0.716667
3         Chris Bosh       2547   27   23  0.851852
4        Dwyane Wade       2548   36   24  0.666667
5        Chris Kaman       2549   17   12  0.705882
6       Kirk Hinrich       2550   22   15  0.681818
7          T.J. Ford       2551   17   15  0.882353
8   Michael Sweetney       2552    6    5  0.833333
9       Jarvis Hayes       2553   11    9  0.818182
10   Mickael Pietrus       2554   17   10  0.588235
11      Marcus Banks       2556   19   14  0.736842
12      Luke Ridnour       2557   10    8  0.800000
13      Reece Gaines       2558    2    1  0.500000
14         Troy Bell       2559    3    3  1.000000
15   Zarko Cabarkapa       2560   14    9  0.642857
16        David West       2561   16   11  0.687500
17    Sasha Pavlovic       2562    6    6  1.000000
18     Dahntay Jones       2563   10    4  0.400000
19        Boris Diaw       2564    8    4  0.500000
20    Zoran Planinic       2565   12    8  0.666667
21     Travis Outlaw       2566    4    2  0.500000
22        Brian Cook       2567   17   14  0.823529
24  Kendrick Perkins       2570    9    6  0.666667
25   Leandro Barbosa       2571    4    2  0.500000
26       Josh Howard       2572    7    3  0.428571
27      Maciej Lampe       2573   11    8  0.727273
28      Jason Kapono       2574    5    5  1.000000
29       Luke Walton       2575    5    4  0.800000
31     Travis Hansen       2580    2    2  1.000000
32       Steve Blake       2581    2    2  1.000000
34      Willie Green       2584   10    7  0.700000
35     Zaza Pachulia       2585    4    3  0.750000
36      Keith Bogans       2586   18    6  0.333333
37       Mo Williams       2590   10    6  0.600000
38       James Jones       2592    2    2  1.000000
39       Kyle Korver       2594    2    2  1.000000
40    Brandon Hunter       2599    3    1  0.333333

In [7]:
df_players_10_games.to_csv("2003_players_10_games.csv", index=True)

Because each player is a binomial and their free throw variance is different we need to apply shrinkage differently to each player.

Reason for applying different shrinkages is beacause if:
Player A:
- has shot 200FT

Player B:
- has shot 20FT

it is not accurate to apply the same shrinkage to both of them

shrinkage equation

$c = 1 - \frac{(k - 3)\sigma^2}{\sum (y - \bar{y})^2}$

$z = \bar{y} + c (y - \bar{y})$

where:
- z is the JS adjusted average
- $\bar{y}$ is the "grand" average


In [8]:
import numpy as np

In [9]:
# applying james stein
# lets try with changing shrinkage c
grand_average = df_players_10_games["ft_pct"].mean()
k = len(df_players_10_games)

def js_avg(df):
    # free throws are a binomial process
	variance = df["ft_pct"] * (1-df["ft_pct"]) / df["fta"]

	shrinkage_c = 1 - ((k-3)*variance) / sum((df["ft_pct"] - grand_average)**2)
	# lowkey apply a sigmoid function here to limit shrinkage c between 0 and 1
	shrinkage_c = 1 / (1 + np.exp(shrinkage_c))
	df["variance"] = variance
	df["shrinkage_c"] = shrinkage_c
	df["js_ft_pct"] = grand_average + shrinkage_c * (df["ft_pct"] - grand_average)

	return df

df_players_10_games_js = js_avg(df_players_10_games)

In [10]:
df_players_10_games_js

player_name  player_id  fta  ftm    ft_pct  variance  shrinkage_c  \
0       LeBron James       2544   38   23  0.605263  0.006287     0.298280   
1      Darko Milicic       2545    3    1  0.333333  0.074074     0.668712   
2    Carmelo Anthony       2546   60   43  0.716667  0.003384     0.284505   
3         Chris Bosh       2547   27   23  0.851852  0.004674     0.290578   
4        Dwyane Wade       2548   36   24  0.666667  0.006173     0.297729   
5        Chris Kaman       2549   17   12  0.705882  0.012212     0.327540   
6       Kirk Hinrich       2550   22   15  0.681818  0.009861     0.315750   
7          T.J. Ford       2551   17   15  0.882353  0.006106     0.297409   
8   Michael Sweetney       2552    6    5  0.833333  0.023148     0.385087   
9       Jarvis Hayes       2553   11    9  0.818182  0.013524     0.334211   
10   Mickael Pietrus       2554   17   10  0.588235  0.014248     0.337925   
11      Marcus Banks       2556   19   14  0.736842  0.010206     0.317463   
12      Luke Ridnour       2557   10    8  0.800000  0.016000     0.346992   
13      Reece Gaines       2558    2    1  0.500000  0.125000     0.866776   
14         Troy Bell       2559    3    3  1.000000  0.000000     0.268941   
15   Zarko Cabarkapa       2560   14    9  0.642857  0.016399     0.349075   
16        David West       2561   16   11  0.687500  0.013428     0.333721   
17    Sasha Pavlovic       2562    6    6  1.000000  0.000000     0.268941   
18     Dahntay Jones       2563   10    4  0.400000  0.024000     0.389733   
19        Boris Diaw       2564    8    4  0.500000  0.031250     0.430009   
20    Zoran Planinic       2565   12    8  0.666667  0.018519     0.360220   
21     Travis Outlaw       2566    4    2  0.500000  0.062500     0.607394   
22        Brian Cook       2567   17   14  0.823529  0.008549     0.309270   
24  Kendrick Perkins       2570    9    6  0.666667  0.024691     0.393519   
25   Leandro Barbosa       2571    4    2  0.500000  0.062500     0.607394   
26       Josh Howard       2572    7    3  0.428571  0.034985     0.451165   
27      Maciej Lampe       2573   11    8  0.727273  0.018032     0.357645   
28      Jason Kapono       2574    5    5  1.000000  0.000000     0.268941   
29       Luke Walton       2575    5    4  0.800000  0.032000     0.434239   
31     Travis Hansen       2580    2    2  1.000000  0.000000     0.268941   
32       Steve Blake       2581    2    2  1.000000  0.000000     0.268941   
34      Willie Green       2584   10    7  0.700000  0.021000     0.373465   
35     Zaza Pachulia       2585    4    3  0.750000  0.046875     0.519310   
36      Keith Bogans       2586   18    6  0.333333  0.012346     0.328214   
37       Mo Williams       2590   10    6  0.600000  0.024000     0.389733   
38       James Jones       2592    2    2  1.000000  0.000000     0.268941   
39       Kyle Korver       2594    2    2  1.000000  0.000000     0.268941   
40    Brandon Hunter       2599    3    1  0.333333  0.074074     0.668712   

    js_ft_pct  
0    0.675069  
1    0.456376  
2    0.708134  
3    0.747488  
4    0.693405  
5    0.705115  
6    0.697503  
7    0.757564  
8    0.754260  
9    0.742654  
10   0.665371  
11   0.714932  
12   0.737795  
13   0.527277  
14   0.784148  
15   0.683139  
16   0.698987  
17   0.784148  
18   0.585973  
19   0.616701  
20   0.691026  
21   0.580382  
22   0.741479  
24   0.689758  
25   0.580382  
26   0.580143  
27   0.712799  
28   0.784148  
29   0.746106  
31   0.784148  
32   0.784148  
34   0.702970  
35   0.728244  
36   0.582840  
37   0.663920  
38   0.784148  
39   0.784148  
40   0.456376

In [57]:
grand_average

np.float64(0.7047410108871327)

In [63]:
k

38

In [67]:
# lets try transforming the data a lil
transform = lambda k, ft_pct: np.sqrt(4*k) * np.arcsin(np.sqrt(ft_pct))
untransform = lambda k, transformed_value: (np.sin(transformed_value / np.sqrt(4*k)))**2

In [70]:
def js_transformed(df):
	x = transform(k, df["ft_pct"])
	grand_average = np.mean(x)
	diff = x - grand_average
	shrinkage_c = 1 - (k-3) / sum(diff**2)
	df["shrinkage_c"] = shrinkage_c

	# untransform
	js_trans_avg = grand_average + shrinkage_c * diff
	df["js_ft_pct"] = untransform(k, js_trans_avg)
	return df

df_players_10_games_js_transformed = js_transformed(df_players_10_games)

In [71]:
df_players_10_games_js_transformed

player_name  player_id  fta  ftm    ft_pct  shrinkage_c  js_ft_pct  \
0       LeBron James       2544   38   23  0.605263     0.928306   0.616359   
1      Darko Milicic       2545    3    1  0.333333     0.928306   0.363020   
2    Carmelo Anthony       2546   60   43  0.716667     0.928306   0.719291   
3         Chris Bosh       2547   27   23  0.851852     0.928306   0.845401   
4        Dwyane Wade       2548   36   24  0.666667     0.928306   0.673062   
5        Chris Kaman       2549   17   12  0.705882     0.928306   0.709311   
6       Kirk Hinrich       2550   22   15  0.681818     0.928306   0.687062   
7          T.J. Ford       2551   17   15  0.882353     0.928306   0.874362   
8   Michael Sweetney       2552    6    5  0.833333     0.928306   0.827950   
9       Jarvis Hayes       2553   11    9  0.818182     0.928306   0.813730   
10   Mickael Pietrus       2554   17   10  0.588235     0.928306   0.600631   
11      Marcus Banks       2556   19   14  0.736842     0.928306   0.737980   
12      Luke Ridnour       2557   10    8  0.800000     0.928306   0.796722   
13      Reece Gaines       2558    2    1  0.500000     0.928306   0.518978   
14         Troy Bell       2559    3    3  1.000000     0.928306   0.998607   
15   Zarko Cabarkapa       2560   14    9  0.642857     0.928306   0.651073   
16        David West       2561   16   11  0.687500     0.928306   0.692313   
17    Sasha Pavlovic       2562    6    6  1.000000     0.928306   0.998607   
18     Dahntay Jones       2563   10    4  0.400000     0.928306   0.425797   
19        Boris Diaw       2564    8    4  0.500000     0.928306   0.518978   
20    Zoran Planinic       2565   12    8  0.666667     0.928306   0.673062   
21     Travis Outlaw       2566    4    2  0.500000     0.928306   0.518978   
22        Brian Cook       2567   17   14  0.823529     0.928306   0.818744   
24  Kendrick Perkins       2570    9    6  0.666667     0.928306   0.673062   
25   Leandro Barbosa       2571    4    2  0.500000     0.928306   0.518978   
26       Josh Howard       2572    7    3  0.428571     0.928306   0.452519   
27      Maciej Lampe       2573   11    8  0.727273     0.928306   0.729112   
28      Jason Kapono       2574    5    5  1.000000     0.928306   0.998607   
29       Luke Walton       2575    5    4  0.800000     0.928306   0.796722   
31     Travis Hansen       2580    2    2  1.000000     0.928306   0.998607   
32       Steve Blake       2581    2    2  1.000000     0.928306   0.998607   
34      Willie Green       2584   10    7  0.700000     0.928306   0.703870   
35     Zaza Pachulia       2585    4    3  0.750000     0.928306   0.750184   
36      Keith Bogans       2586   18    6  0.333333     0.928306   0.363020   
37       Mo Williams       2590   10    6  0.600000     0.928306   0.611498   
38       James Jones       2592    2    2  1.000000     0.928306   0.998607   
39       Kyle Korver       2594    2    2  1.000000     0.928306   0.998607   
40    Brandon Hunter       2599    3    1  0.333333     0.928306   0.363020   

    variance  
0   0.006287  
1   0.074074  
2   0.003384  
3   0.004674  
4   0.006173  
5   0.012212  
6   0.009861  
7   0.006106  
8   0.023148  
9   0.013524  
10  0.014248  
11  0.010206  
12  0.016000  
13  0.125000  
14  0.000000  
15  0.016399  
16  0.013428  
17  0.000000  
18  0.024000  
19  0.031250  
20  0.018519  
21  0.062500  
22  0.008549  
24  0.024691  
25  0.062500  
26  0.034985  
27  0.018032  
28  0.000000  
29  0.032000  
31  0.000000  
32  0.000000  
34  0.021000  
35  0.046875  
36  0.012346  
37  0.024000  
38  0.000000  
39  0.000000  
40  0.074074

In [78]:
grand_average

np.float64(0.7047410108871327)

im stuck.

i want the shrinkage applied to each player to be different since variances across players differ depending on the amount of free throws they take.

1. ive tried the appraoch where i calculate individual variances for players but the shrinkage becomes wild especially for players with smaller sample sizes. For example someone who is a 33% ft shooter with 3 attempts gets shrunk to 97%. players with 100% ft% remain at 100%

2. ive tried another approach where i assume constant variance across all players. shrinkage is too powerful and everyone is assumed to shoot at the grand average.

3. ive also tried arcsin transformation. shrinkage is negligible and it also assumes constant variance. 

i think the approach im gonna try next is a modification approach of 1. where i apply a weighting factor on the variance. should look at existing approaches for this